keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1061,28 +1062,91 @@ Details to consider:
-repeated evaluations if they are shared across validations.
+(cel-policy-template does this).
+
+A policy can include an additional `variables` section. This is an array
+containing one or more `name` and `expression` pairs, which can be used/re-used by
+the policy's validation expressions. These results are memoized on a
+per-validation basis, so if multiple expressions use the same spec variables,
+the expression that calculates the variable's value will only be ran once.
+
+The variables can be accessed as members of `variables`, which is an object
+that is exposed to CEL expressions (both validation expressions as well as
+other variables).
+
+For example:
 
 ```yaml
   variables:
     - name: metadataList
       expression: "spec.list.map(x, x.metadata)"
     - name: itemMetadataNames
-      expression: "metadataList.map(m, m.name)"
+      expression: "variables.metadataList.map(m, m.name)"
   validations:
-    - expression: "itemMetadataNames.all(name, name.startsWith('xyz-'))"
-    - expression: "itemMetadataNames.exists(name, name == 'required')"
+    - expression: "variables.itemMetadataNames.all(name, name.startsWith('xyz-'))"
+    - expression: "variables.itemMetadataNames.exists(name, name == 'required')"
 ```
 
+Variable names must be valid CEL names, and not already used by Kubernetes
+in some way (i.e, `request` is not considered valid). What constitutes a
+valid CEL name can be found at CEL's [language definition](https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax) under `IDENT`.
+This validity is checked at creation time.
+
+For runtime cost limit purposes, variables count towards the runtime cost limit
+once per policy. The cost of each variable is computed when it is first evaluated in an
+expression, mirroring how the cost limit would be calculated if the variable's
+expression was embedded verbatim.  If the runtime cost limit is exceeded in the
+process, then evaluation halts. Whether or not the request actually fails
+depends on the failure policy, however. For subsequent uses, inclusion of the
+variable has zero effect on the runtime cost limit. If the variable evaluates
+to an array or some other iterable, and some expression iterates on it, that
+of course contributes to the cost limit, but simply including the variable does
+not add the underlying expression's cost again.
+
+Variables must evaluate at compile time to a CEL type that is not the base
+`ref.Val`. This also means that unlike policy expressions, variables have
+compile-time type checking performed on them (since types are all known).
+A compile-time type check failure prevents the policy from being created,
+regardless of failure policy.
+
+Variables can reference other variables, but circular references
+are not allowed. Variables can also only reference other variables that
+have been previously defined in the `variables` section.
+
+If an error ocurrs during variable evaluation, then the expression
```

Comment:
Does this mean errors are never cached? For a malformed expression it may evaluate it over and over again.

Reply:
Isn't it possible that an expression fails for some inputs and not others, and those inputs could change from one evaluation to the other? I think we could cache if inputs are a factor in cache invalidation, but I'm not sure if we get something out of that considering how fast expression evaluation is intended to be.

Reply:
Sorry for the confusion. Since the variables are not cached between evaluations, I think the problem above would not apply because an error of variable evaluation will fast fail the whole validation.

Reply:
(resolved)